core/blockchain.go

```diff
@@ -2124,6 +2128,43 @@ func (bc *BlockChain) WriteShardLastCrossLink(shardID uint32, cl types.CrossLink
 	return rawdb.WriteShardLastCrossLink(bc.db, cl.ShardID(), cl.Serialize())
 }
 
+// LastContinuousCrossLink saves the last crosslink of a shard
+// This function will update the latest crosslink in the sense that
+// any previous block's crosslink is received up to this point
+// there is no missing hole between genesis to this crosslink of given shardID
+func (bc *BlockChain) LastContinuousCrossLink(shardID uint32, cl types.CrossLink) error {
+	cl0, err := bc.ReadShardLastCrossLink(shardID)
+	if cl0 == nil {
+		bc.WriteShardLastCrossLink(shardID, cl)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	newCheckpoint := uint64(0)
+	for i := cl0.BlockNum() + 1; i < cl.BlockNum(); i++ {
+		tmp, err := bc.ReadShardLastCrossLink(shardID)
+		if err != nil || tmp == nil {
+			if i > cl0.BlockNum() {
+				newCheckpoint = i - 1
```

Comment:
Should you break here? Here you found a new checkPoint right?

Reply:
yes, that's right